woocommerce/woocommerce
# Prevent unnecessary admin feature class loading and initialization on frontend

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Closes https://github.com/woocommerce/woocommerce/issues/38308

This PR optimizes how WooCommerce admin features are loaded by preventing unnecessary initialization on frontend requests.

- Prevents unnecessary class autoloading on frontend requests
- Maintains full functionality for admin, AJAX, cron, CLI, and REST API context, except for store API requests. I checked the code, and from my understanding, store API requests are designed for frontend usage and don't need to load admin features. Please correct me if I'm wrong.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a fresh site with `wc-smooth-generator` plugin installed and activated.
2. Complete the core profiler
3. Confirm the coming soon footer banner is shown when the site is coming soon for admin users
4. Test around the admin pages and confirm no errors
5. Go to WooCommerce Admin Test Helper (Tools > WCA Test Helper) > Tools
6. Run `wc_admin_daily` and confirm it runs successfully
7. Go to WooCommerce Admin Test Helper >  `Features` and confirm you can toggle the features
8. Go to Tools > Smooth Generator > Generate Products and confirm no errors
9. Add the following code to `Features.php` before [line 56](https://github.com/woocommerce/woocommerce/blob/eb4cfbcc4e5b10780d72bd7097e67b40c9d37afa/plugins/woocommerce/src/Admin/Features/Features.php#L55):

```php
spl_autoload_register(
	function( $classname ) {
		error_log( print_r( $classname, true ) );
	}
);
```

10. Add `define( 'DISABLE_WP_CRON', true );` to `wp-config.php` to prevent wp-cron from loading admin features by itself
11. Go to frontend in incognito mode and confirm no errors
12. Confirm `wp-content/debug.log` doesn't log any `Automattic\WooCommerce\Admin\Features\` classes




### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
